### PR TITLE
CP-31450: Add domid to Datapath.attach

### PR DESCRIPTION
**Pull request by BenSimsCitrix**
We now honour the existing API of Datapath.attach by
passing the domid of the vm that the vbd is being attached to.

This will enable a qemu process per vm.

Signed-off-by: ben sims <ben.sims@citrix.com>

### CODE REVIEW

**Comment by edwintorok** [comment]
The documentation in xapi-storage currently says that `domain` is an "An opaque string which represents the Xen domain.". If we start using it as a domid it is not opaque anymore, so we should also update this documentation.
If we know this is always going to be a domain ID, then perhaps we can pass an integer here, and turn it into the string just before calling the SMAPIv3 API. (changing SMAPIv3 from string to int is probably not doable without backwards compatibility problems)

**Comment by BenSimsCitrix** [comment]
The iso as whole builds successfuly on   private/bensi/CP-31450/12

I observe in the logs that the domid passed to Datapath.attach is now correct and consistent with the output of xl list.

I'd like to test this more thoroughly, so im keen to hear you ideas.

**The following pull requests need to be simultaneous otherwise the build will break** 

https://github.com/xapi-project/xcp-idl/pull/278
https://github.com/xapi-project/xen-api/pull/3910
https://github.com/xapi-project/xapi-storage-script/pull/85
https://github.com/xapi-project/xenopsd/pull/644

File: storage/storage_interface.ml [hunk 1]
```diff
@@ -995,7 +1005,7 @@ sig
     val attach2 :
       context ->
       dbg: debug_info ->
-      dp: dp -> sr: sr -> vdi: vdi -> read_write: bool -> backend
+      dp: dp -> sr: sr -> vdi: vdi -> vm: vm -> read_write: bool -> backend
```
Comment (edwintorok) [inline storage/storage_interface.ml]: @BenSimsCitrix raised a good point that we cannot change this interface here due to storage migration from older versions. Even having it as optional wouldn't help if the new storage plugin actually requires a correct value here.
So the solution would be to have an `attach3`, `activate3`, `deactivate3`, `detach3` impl.

Comment (BenSimsCitrix) [reply storage/storage_interface.ml]: I've added new apis attach3 and activate3, to preserve backwards compatability 
Ive also run the following tests

DiffHostCrossPoolSXM Job ID 2594732 (Old version Lima)
DiffHostCrossPoolSXM Job ID 2594733 (Old version Dundee)

**Comment by MarkSymsCtx** [comment]
> The documentation in xapi-storage currently says that `domain` is an "An opaque string which represents the Xen domain.". If we start using it as a domid it is not opaque anymore, so we should also update this documentation.
> If we know this is always going to be a domain ID, then perhaps we can pass an integer here, and turn it into the string just before calling the SMAPIv3 API. (changing SMAPIv3 from string to int is probably not doable without backwards compatibility problems).
> Or is this going to be the domain's UUID, in which case it is a string?

**Comment by TimSmithCtx** [comment]
> The documentation in xapi-storage currently says that `domain` is an "An opaque string which represents the Xen domain.". If we start using it as a domid it is not opaque anymore, so we should also update this documentation.
> If we know this is always going to be a domain ID, then perhaps we can pass an integer here, and turn it into the string just before calling the SMAPIv3 API. (changing SMAPIv3 from string to int is probably not doable without backwards compatibility problems).
> Or is this going to be the domain's UUID, in which case it is a string?

We absolutely have to have the domain ID from somewhere as we have to attach to it. The current mechanism is rather magic-behind-the-scenes.

However we also want to support multiple service instances in Dom0 as we might want to give them different I/O weights. So it's more the case that *if* it's numeric it's the domain ID, and if it isn't (or is 0) then it's an identifier for a Dom0 service which is separate from other Dom0 services

**Comment by BenSimsCitrix** [comment]
This has passed 

Storage BST Job ID 110188 
DiffHostCrossPoolSXM Job ID 2594732 (Old version Lima)
DiffHostCrossPoolSXM Job ID 2594733 (Old version Dundee)

Latest changes have also built here.

https://ratchet.do.citrite.net/job/xenserver-build/job/private%252Fbensi%252FCA-312665/

**Comment by BenSimsCitrix** [comment]
This has passed

Storage BST Job ID 110188
DiffHostCrossPoolSXM Job ID 2594732 (Old version Lima)
DiffHostCrossPoolSXM Job ID 2594733 (Old version Dundee)

Latest changes have also built here.

https://ratchet.do.citrite.net/job/xenserver-build/job/private%252Fbensi%252FCA-312665/

**Comment by edwintorok** [comment]
Looks like the unit tests are failing here

File: storage/storage_interface.ml [hunk 2]
```diff
@@ -113,6 +113,15 @@ end
 type vdi = Vdi.t
 let vdi_pp : Format.formatter -> vdi -> unit = fun ppf vdi -> Format.fprintf ppf "%s" (Vdi.string_of vdi)
 
+module Vm : WRAPPEDSTRING = struct
+  type t = string [@@deriving rpcty]
```
Comment (edwintorok) [inline storage/storage_interface.ml]: might be worth adding a comment here that this is actually a `domid`, at least for now.

**Comment by BenSimsCitrix** [comment]
This is ready for merge unless there are more review issues

**Pull request merged**